### PR TITLE
Performance: Optimize checklist rendering

### LINKED ISSUE #8446: Performance: Optimize checklist rendering
## Feature Description

Checklist currently renders the list even if it's closed.
For example page preview/thumbnail is rendered for "text too little" warning as soon as you add something to the page (and on many pages in real-world stories).
Then the list is re-rendered with some basic actions like DND (affects #8259).
Thumbnail rendering will be improved with #7987 but I think it still shouldn't render checklist contents if they are not visible.

image

image

## Additional Context

---

_Do not alter or remove anything below. The following sections will be managed by moderators only._

### PR DESCRIPTION
**Pull request by maxyinger**
## Context
Part of tightening up perf of the editor for primary actions like DnD

## Summary
Removes unneeded rendering of elements within the checklist when the checklist is closed

Screen Shot 2021-07-30 at 1 18 58 PM
Screen Shot 2021-07-30 at 1 18 42 PM

## Relevant Technical Choices
just used a provider and transition group callbacks to see when the component was mounted and pass that mounted value around.

## To-do
NA

## User-facing changes
NA

## Testing Instructions
Open up the editor help center and see that there were no regressions. Can inspect react flame chart and see that it takes up less time to render when it's closed as it did before or just check rendered dom elements and see that there are no rendered children of the popup when it's closed.

### QA

- [x] This is a non-user-facing change and requires no QA

This PR can be tested by following these steps:

1.

### UAT

- [x] UAT should use the same steps as above.

This PR can be tested by following these steps:

1.

## Checklist

- [x] This PR addresses an existing issue and I have linked this PR to it in ZenHub
- [x] I have tested this code to the best of my abilities
- [x] I have verified accessibility to the best of my abilities ([docs](https://github.com/google/web-stories-wp/blob/main/docs/accessibility-testing.md))
- [x] I have verified i18n and l10n (translation, right-to-left layout) to the best of my abilities
- [ ] This PR contains automated tests (unit, integration, and/or e2e) to verify the code works as intended ([docs](https://github.com/google/web-stories-wp/tree/main/docs#testing))
- [x] I have added documentation where necessary
- [x] I have added a matching `Type: XYZ` label to the PR

---

Fixes #8446

### CODE REVIEW

File: assets/src/edit-story/components/checklist/checklist.js [hunk 1]
```diff
@@ -60,6 +62,53 @@ const Wrapper = styled.div`
...
+    <Popup
+      popupId={POPUP_ID}
+      isOpen={isOpen}
+      ariaLabel={CHECKLIST_TITLE}
+      shouldKeepMounted
+      onEnter={() => setIsMounted(true)}
+      onExited={() => setIsMounted(false)}
+    >
+      <mountedContext.Provider value={isMounted}>
+        {isMounted ? (
```
Comment (dreamofabear) [inline assets/src/edit-story/components/checklist/checklist.js]: Naive question: why can't we use `isOpen` here?

Comment (maxyinger) [reply assets/src/edit-story/components/checklist/checklist.js]: not naive at all. that was my original approach, but it causes children to get unmounted while the exit animation is still playing which results in some unwanted layout and nasty transition.

using `isMounted` , which looks at the transition callbacks, allows this to wait until the exit animation is complete before removing the children components

Comment (dreamofabear) [reply assets/src/edit-story/components/checklist/checklist.js]: Thanks! Any reason why we should create a new context vs. adding `isMounted` to `useChecklist`?

Comment (maxyinger) [reply assets/src/edit-story/components/checklist/checklist.js]: nope. good call out. happy to move it in there to consolidate it a bit

**Review by BrittanyIRL — APPROVED** [review]
Tested locally and works well!  

Running on my machine, the flamegraph for this update vs how main is right now  I get the below (main is on the left) - I loaded a new story on each and then toggled the checklist open closed, added a title with the checklist closed, opened it, added some pages (i think i added one more on main, thus the 2 extra rendered lines) 

Screen Shot 2021-08-02 at 9 56 17 AM

The difference in time isn't too big but it _feels_ faster. 

Most of the reasons for the first like 6 renders is due to upsteram story providers: 
Screen Shot 2021-08-02 at 9 58 58 AM

I'm curious if we should make another ticket to explore ways to cut down how this causes the checklist to re-render. Thoughts? 

I agree with Will to consolidate this new bit of context. 

Small nit/question - we would only really need to check if `isChecklistMounted` on the checks if `isRendered` is true, is there any value in checking that first and then if it's true checking `isChecklistMounted`?

**Comment by maxyinger** [comment]
> I'm curious if we should make another ticket to explore ways to cut down how this causes the checklist to re-render. Thoughts?

I'm wondering if this is causing upstream rerenders? There may be similar things like it in other upstream providers. We can always look into it more if needed in this and other providers. Can at least clean this up in this PR if wanted
https://github.com/google/web-stories-wp/blob/4f0be63f3e61b47a92023b81ec338d54ea3ef0fe/assets/src/edit-story/components/checklist/checkpointContext/checkpointContext.js#L79-L87

> I agree with Will to consolidate this new bit of context.

👍 

> Small nit/question - we would only really need to check if isChecklistMounted on the checks if isRendered is true, is there any value in checking that first and then if it's true checking isChecklistMounted?

My instincts tell me there wouldn't be much value. Since we can't conditionally call hooks, we'd need to move that check into a child component and I feel like another child component in the react tree would be more overhead than just pulling a boolean from context. Not 100% sure here, but I feel like if there is a difference either way, it'd be negligible 🤷

**Comment by BrittanyIRL** [comment]
>  it'd be negligible

Cool 😎  

🚀

**Comment by BrittanyIRL** [comment]
> I'm wondering if this is causing upstream rerenders? There may be similar things like it in other upstream providers. We can always look into it more if needed in this and other providers. Can at least clean this up in this PR if wanted

hmmmm 🤔   

maybe worth timeboxing and seeing if it helps? I don't want to add scope here, so if it ends up complicating things let's do it on it's own.

**Comment by maxyinger** [comment]
> maybe worth timeboxing and seeing if it helps? I don't want to add scope here, so if it ends up complicating things let's do it on it's own.

I can at least clean up this `useStory` mentioned here and it definitely won't hurt anything. won't add any time to the PR either

**Pull request merged**